### ISSUE TITLE
Visual Dataset in memory mode(driver='core') got empty results(full of zeros) when use create_virtual_dataset

### ISSUE THREAD

**Issue by xmurobi** [h5py/h5py#1805]
To assist reproducing bugs, please include the following:
 * Operating System (e.g. Windows 10, MacOS 10.11, Ubuntu 16.04.2 LTS, CentOS 7)
    macOS High Sierra 10.13.6
 * Python version (e.g. 2.7, 3.5)
   Python 3.7.6 (default, Dec 30 2019, 19:38:36) 
   [Clang 10.0.0 (clang-1000.11.45.5)] on darwin
 * Where Python was acquired (e.g. system Python on MacOS or Linux, Anaconda on
   Windows)
 * h5py version (e.g. 2.6)
   3.1.0
 * HDF5 version (e.g. 1.8.17)
   1.12
 * The full traceback/stack trace shown (if it appears)

```
import h5py
import numpy as np

dpath = "/a/data"
dtype = [("ID", "i8"), ("PRICE", "f8")]
# create some sample data
# data = np.arange(0, 101).reshape(1, 101) + np.arange(1, 5).reshape(4, 1)
data = [np.array([(j+i,j+i+0.1) for j in range(100)], dtype=dtype) for i in range(1,6)]
fs = []
ds = []

# Create source files (0.h5 to 3.h5)
for n in range(5):
    f = h5py.File(f"{n}.h5", "w", libver="latest")
    f.swmr_mode = True
    d = f.create_dataset(dpath, (100, ), dtype, data[n], maxshape=(None, ), chunks=True)
    fs.append(f)
    ds.append(d)

# uncomment this everything goes right #
# for f in fs:
    # f.close()
    
layout = h5py.VirtualLayout(shape=(400,), dtype=dtype, maxshape=(None,))
for n in range(4):
    filename = "{}.h5".format(n)
    vsource = h5py.VirtualSource(filename, dpath, shape=(100,))
    layout[n*100:(n+1)*100] = vsource[:]


# Add virtual dataset to output file
f = h5py.File("vds.h5", "w", libver="latest", driver='core', backing_store=False, swmr=True)
f.create_virtual_dataset("vdata", layout)
f.create_dataset(dpath, data=data, dtype=dtype)
print(f["vdata"][:])
f.close()

```

Run above code got array:

[(0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.)
 (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.)
 (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.)
 (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.)
 (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.)
 (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.) (0, 0.)
...

But expect

[(  1,   1.1) (  2,   2.1) (  3,   3.1) (  4,   4.1) (  5,   5.1)
 (  6,   6.1) (  7,   7.1) (  8,   8.1) (  9,   9.1) ( 10,  10.1)
 ( 11,  11.1) ( 12,  12.1) ( 13,  13.1) ( 14,  14.1) ( 15,  15.1)
 ( 16,  16.1) ( 17,  17.1) ( 18,  18.1) ( 19,  19.1) ( 20,  20.1)
 ( 21,  21.1) ( 22,  22.1) ( 23,  23.1) ( 24,  24.1) ( 25,  25.1)
 ( 26,  26.1) ( 27,  27.1) ( 28,  28.1) ( 29,  29.1) ( 30,  30.1)
...

Only when I uncomment  lines:
```
for f in fs:
    f.close()
```
before visual dataset creation will make things goes right.

But actually in my case, all files should not be closed because the coming datastream will not stop and I want to make some memory visual datasets to tracking those h5 files.

**Comment by jangarrevoet** [comment]
You will have to explicitly flush the data in your writer. Otherwise the hdf5lib will determine when to write to the file.

**Comment by xmurobi** [comment]
@jangarrevoet 

call f.flush() doesn't work...

**Comment by xmurobi** [comment]
I found that after change the line below

`f = h5py.File("vds.h5", "w", libver="latest", driver='core', backing_store=False, swmr=True)`

to

`f = h5py.File("vds.h5", "w", libver="latest")`

Then everything works fine. No need to close the files above at all.

Why？

**Comment by xmurobi** [comment]
After do some investigation I found:

If I call:

The code from function **close** in h5py/_hl/files.py
```

                # We have to explicitly murder all open objects related to the file

                # Close file-resident objects first, then the files.
                # Otherwise we get errors in MPI mode.
                id_list = h5f.get_obj_ids(self.id, ~h5f.OBJ_FILE)
                file_list = h5f.get_obj_ids(self.id, h5f.OBJ_FILE)

                id_list = [x for x in id_list if h5i.get_file_id(x).id == self.id.id]
                file_list = [x for x in file_list if h5i.get_file_id(x).id == self.id.id]

                for id_ in id_list:
                    while id_.valid:
                        h5i.dec_ref(id_)

                for id_ in file_list:
                    while id_.valid:
                        h5i.dec_ref(id_)

```
Then: 

`print(f["vdata"][:])`

That works!

So, could I simply wrapper a function which use same way to replace the original close and solve this issue temporarly @takluyver?

**Comment by takluyver** [comment]
Have you tried calling `d.flush()` (on the datasets) as well as `f.flush()`?

One of the most annoying things about virtual datasets is that any error accessing the underlying data is silenced and it just returns 'empty' data instead, hiding what the real problem is. I would guess that in this case how SWMR & virtual datasets interact is not well defined.

**Comment by xmurobi** [comment]
> Have you tried calling `d.flush()` (on the datasets) as well as `f.flush()`?
> 
> One of the most annoying things about virtual datasets is that any error accessing the underlying data is silenced and it just returns 'empty' data instead, hiding what the real problem is. I would guess that in this case how SWMR & virtual datasets interact is not well defined.

@takluyver Yes, I do call `d.flush()` in code but doesn't work. I had made a multiple processes version to test this and the result seems the same. Basically speaking, when the file opened for SWMR mode and holded by other process, virtual dataset won't work until the file been closed.

**Comment by xmurobi** [comment]
Hi, @takluyver I found the line 787 in h5py/_hl/dataset.py is the key:

`self.id.read(mspace, fspace, arr, mtype, dxpl=self._dxpl)`

I had set a breakpoint on this line and manually close all files that vds linked to which opened in another process use swmr writer mode, then this line works, the arr got the data. Otherwise, the arr got full zero results.

But for a normal dataset, which opened as swmr read mode in a separated process, this line works fine and always got data.

Is that a bug? Or VDS only designed for _cold_ data?

**Comment by takluyver** [comment]
> Is that a bug? Or VDS only designed for cold data?

It's probably a bug, but VDS and SMWR features were added at about the same time for different use cases, so I'm not that surprised that there are some problems when they're used together.

If there is a bug, it's probably in HDF5 itself. h5py doesn't do anything special to read virtual datasets, it makes the same call into HDF5 (the H5Dread function) as it does for any dataset. You could try asking help@hdfgroup.org, or on the [HDF forum](https://forum.hdfgroup.org/). But they may want someone to reproduce it using C code to confirm that it's an issue with HDF5 itself.
